diff --git a/src/index/base.cpp b/src/index/base.cpp
index 8fe30f8960..8f8054b329 100644
--- a/src/index/base.cpp
+++ b/src/index/base.cpp
@@ -105,7 +105,7 @@ bool BaseIndex::Init()
             }
         }
         if (prune_violation) {
-            return InitError(strprintf(Untranslated("%s best block of the index goes beyond pruned data. Please disable the index or reindex (which will download the whole blockchain again)"), GetName()));
+            return InitError(strprintf(Untranslated("%s best block of the index goes beyond pruned data. Please disable the index or reindex (which will download the whole blockchain again)"), GetName())); // peercoin: should never happen
         }
     }
     return true;
