diff --git a/src/index/txindex.h b/src/index/txindex.h
index 2bbc602631..87f87c9e44 100644
--- a/src/index/txindex.h
+++ b/src/index/txindex.h
@@ -6,6 +6,8 @@
 #define BITCOIN_INDEX_TXINDEX_H
 
 #include <index/base.h>
+#include <index/disktxpos.h>
+#include <primitives/block.h>
 
 /**
  * TxIndex is used to look up transactions included in the blockchain by hash.
@@ -41,6 +43,9 @@ public:
     /// @param[out]  tx  The transaction itself.
     /// @return  true if transaction is found, false otherwise
     bool FindTx(const uint256& tx_hash, uint256& block_hash, CTransactionRef& tx) const;
+
+    bool FindTxPosition(const uint256& txid, CDiskTxPos& pos) const;
+    std::map<uint256,std::pair<CBlockHeader,CTransactionRef>> cachedTxs;
 };
 
 /// The global transaction index, used in GetTransaction. May be null.
