diff --git a/src/script/script.cpp b/src/script/script.cpp
index 9a6419088b..211b57e40d 100644
--- a/src/script/script.cpp
+++ b/src/script/script.cpp
@@ -1,12 +1,13 @@
 // Copyright (c) 2009-2010 Satoshi Nakamoto
 // Copyright (c) 2009-2020 The Bitcoin Core developers
 // Distributed under the MIT software license, see the accompanying
 // file COPYING or http://www.opensource.org/licenses/mit-license.php.
 
 #include <script/script.h>
 
 #include <util/strencodings.h>
 
+#include <key.h>
 #include <string>
 
 std::string GetOpName(opcodetype opcode)
@@ -280,6 +281,19 @@ bool CScript::HasValidOps() const
     return true;
 }
 
+void CScript::SetMultisig(int nRequired, const std::vector<CPubKey>& keys)
+{
+    this->clear();
+
+    *this << EncodeOP_N(nRequired);
+    for (const auto& pubkey : keys)
+    {
+        std::vector<unsigned char> vchPubKey(pubkey.begin(), pubkey.end());
+        *this << vchPubKey;
+    }
+    *this << EncodeOP_N(keys.size()) << OP_CHECKMULTISIG;
+}
+
 bool GetScriptOp(CScriptBase::const_iterator& pc, CScriptBase::const_iterator end, opcodetype& opcodeRet, std::vector<unsigned char>* pvchRet)
 {
     opcodeRet = OP_INVALIDOPCODE;
