diff --git a/src/util/error.cpp b/src/util/error.cpp
index af8cbd0353..898739b5ba 100644
--- a/src/util/error.cpp
+++ b/src/util/error.cpp
@@ -29,8 +29,6 @@ bilingual_str TransactionErrorString(const TransactionError err)
             return Untranslated("PSBTs not compatible (different transactions)");
         case TransactionError::SIGHASH_MISMATCH:
             return Untranslated("Specified sighash value does not match value stored in PSBT");
-        case TransactionError::MAX_FEE_EXCEEDED:
-            return Untranslated("Fee exceeds maximum configured by user (e.g. -maxtxfee, maxfeerate)");
         case TransactionError::EXTERNAL_SIGNER_NOT_FOUND:
             return Untranslated("External signer not found");
         case TransactionError::EXTERNAL_SIGNER_FAILED:
