diff --git a/src/script/interpreter.cpp b/src/script/interpreter.cpp
index 11b1a1c887..8ab82e4650 100644
--- a/src/script/interpreter.cpp
+++ b/src/script/interpreter.cpp
@@ -1348,6 +1348,10 @@ public:
     void Serialize(S &s) const {
         // Serialize nVersion
         ::Serialize(s, txTo.nVersion);
+        if (txTo.nVersion < 3) {
+            // Serialize nTime
+            ::Serialize(s, txTo.nTime);
+        }
         // Serialize vin
         unsigned int nInputs = fAnyoneCanPay ? 1 : txTo.vin.size();
         ::WriteCompactSize(s, nInputs);
@@ -1620,6 +1624,10 @@ uint256 SignatureHash(const CScript& scriptCode, const T& txTo, unsigned int nIn
         CHashWriter ss(SER_GETHASH, 0);
         // Version
         ss << txTo.nVersion;
+        if (txTo.nVersion < 3) {
+            // nTime
+            ss << txTo.nTime;
+        }
         // Input prevouts/nSequence (none/all, depending on flags)
         ss << hashPrevouts;
         ss << hashSequence;
