diff --git a/src/script/sign.cpp b/src/script/sign.cpp
index 371a937bc8..287c318b18 100644
--- a/src/script/sign.cpp
+++ b/src/script/sign.cpp
@@ -622,6 +622,10 @@ bool SignTransaction(CMutableTransaction& mtx, const SigningProvider* keystore,
 {
     bool fHashSingle = ((nHashType & ~SIGHASH_ANYONECANPAY) == SIGHASH_SINGLE);
 
+    // we don't need nTime anymore
+    if (mtx.nVersion >= 3)
+        mtx.nTime = 0;
+
     // Use CTransaction for the constant parts of the
     // transaction to avoid rehashing.
     const CTransaction txConst(mtx);
