diff --git a/src/script/standard.h b/src/script/standard.h
index eb50421768..e55c80be0e 100644
--- a/src/script/standard.h
+++ b/src/script/standard.h
@@ -32,10 +32,10 @@ public:
 };
 
 /**
- * Default setting for nMaxDatacarrierBytes. 80 bytes of data, +1 for OP_RETURN,
+ * Default setting for nMaxDatacarrierBytes. 256 bytes of data, +1 for OP_RETURN,
  * +2 for the pushdata opcodes.
  */
-static const unsigned int MAX_OP_RETURN_RELAY = 83;
+static const unsigned int MAX_OP_RETURN_RELAY = 259;
 
 /**
  * A data carrying output is an unspendable output containing data. The script
