diff --git a/src/policy/settings.h b/src/policy/settings.h
index 0b4fc1e770..2488ef7530 100644
--- a/src/policy/settings.h
+++ b/src/policy/settings.h
@@ -8,18 +8,15 @@
 
 #include <policy/policy.h>
 
-class CFeeRate;
 class CTransaction;
 
 // Policy settings which are configurable at runtime.
-extern CFeeRate incrementalRelayFee;
-extern CFeeRate dustRelayFee;
 extern unsigned int nBytesPerSigOp;
 extern bool fIsBareMultisigStd;
 
 static inline bool IsStandardTx(const CTransaction& tx, std::string& reason)
 {
-    return IsStandardTx(tx, ::fIsBareMultisigStd, ::dustRelayFee, reason);
+    return IsStandardTx(tx, ::fIsBareMultisigStd, reason);
 }
 
 static inline int64_t GetVirtualTransactionSize(int64_t weight, int64_t sigop_cost)
