diff --git a/src/protocol.cpp b/src/protocol.cpp
index 139405170b..0ec9ad77a7 100644
--- a/src/protocol.cpp
+++ b/src/protocol.cpp
@@ -179,10 +179,11 @@ std::string CInv::ToString() const
 const std::vector<std::string> &getAllNetMessageTypes()
 {
     return allNetMessageTypesVec;
 }
 
+const unsigned int POW_HEADER_COOLING = 70;
 /**
  * Convert a service flag (NODE_*) to a human readable string.
  * It supports unknown service flags which will be returned as "UNKNOWN[...]".
  * @param[in] bit the service flag is calculated as (1 << bit)
  */
