diff --git a/src/util/error.h b/src/util/error.h
index 4cc35eb1fd..98102026bf 100644
--- a/src/util/error.h
+++ b/src/util/error.h
@@ -22,14 +22,13 @@ struct bilingual_str;
 enum class TransactionError {
     OK, //!< No error
     MISSING_INPUTS,
     ALREADY_IN_CHAIN,
     P2P_DISABLED,
     MEMPOOL_REJECTED,
     MEMPOOL_ERROR,
     INVALID_PSBT,
     PSBT_MISMATCH,
     SIGHASH_MISMATCH,
-    MAX_FEE_EXCEEDED,
     EXTERNAL_SIGNER_NOT_FOUND,
     EXTERNAL_SIGNER_FAILED,
 };
