diff --git a/src/util/moneystr.cpp b/src/util/moneystr.cpp
index 2cd7a426f8..17255ee0e2 100644
--- a/src/util/moneystr.cpp
+++ b/src/util/moneystr.cpp
@@ -15,24 +15,24 @@
 std::string FormatMoney(const CAmount n)
 {
     // Note: not using straight sprintf here because we do NOT want
     // localized number formatting.
     static_assert(COIN > 1);
     int64_t quotient = n / COIN;
     int64_t remainder = n % COIN;
     if (n < 0) {
         quotient = -quotient;
         remainder = -remainder;
     }
-    std::string str = strprintf("%d.%08d", quotient, remainder);
+    std::string str = strprintf("%d.%06d", quotient, remainder);
 
     // Right-trim excess zeros before the decimal point:
     int nTrim = 0;
     for (int i = str.size()-1; (str[i] == '0' && IsDigit(str[i-2])); --i)
         ++nTrim;
     if (nTrim)
         str.erase(str.size()-nTrim, nTrim);
 
     if (n < 0)
         str.insert((unsigned int)0, 1, '-');
     return str;
 }
