diff --git a/src/policy/settings.cpp b/src/policy/settings.cpp
index eb2ec56850..6b15d31d68 100644
--- a/src/policy/settings.cpp
+++ b/src/policy/settings.cpp
@@ -1,14 +1,11 @@
 // Copyright (c) 2009-2010 Satoshi Nakamoto
 // Copyright (c) 2009-2019 The Bitcoin Core developers
 // Distributed under the MIT software license, see the accompanying
 // file COPYING or http://www.opensource.org/licenses/mit-license.php.
 
 #include <policy/settings.h>
 
-#include <policy/feerate.h>
 #include <policy/policy.h>
 
 bool fIsBareMultisigStd = DEFAULT_PERMIT_BAREMULTISIG;
-CFeeRate incrementalRelayFee = CFeeRate(DEFAULT_INCREMENTAL_RELAY_FEE);
-CFeeRate dustRelayFee = CFeeRate(DUST_RELAY_TX_FEE);
 unsigned int nBytesPerSigOp = DEFAULT_BYTES_PER_SIGOP;
