diff --git a/src/protocol.h b/src/protocol.h
index fdeaa9a9c5..9ab533ca03 100644
--- a/src/protocol.h
+++ b/src/protocol.h
@@ -273,7 +273,7 @@ enum ServiceFlags : uint64_t {
     // Nothing
     NODE_NONE = 0,
     // NODE_NETWORK means that the node is capable of serving the complete block chain. It is currently
-    // set by all Bitcoin Core non pruned nodes, and is unset by SPV clients or other light clients.
+    // set by all Bitcoin Core nodes, and is unset by SPV clients or other light clients.
     NODE_NETWORK = (1 << 0),
     // NODE_BLOOM means the node is capable and willing to handle bloom-filtered connections.
     // Bitcoin Core nodes used to support this by default, without advertising this bit,
@@ -511,4 +511,7 @@ public:
 /** Convert a TX/WITNESS_TX/WTX CInv to a GenTxid. */
 GenTxid ToGenTxid(const CInv& inv);
 
+/** peercoin: How much temperature a PoW header will remove */
+extern const unsigned int POW_HEADER_COOLING;
+
 #endif // BITCOIN_PROTOCOL_H
