diff --git a/src/undo.h b/src/undo.h
index a98f046735..a7da14d171 100644
--- a/src/undo.h
+++ b/src/undo.h
@@ -24,7 +24,8 @@ struct TxInUndoFormatter
 {
     template<typename Stream>
     void Ser(Stream &s, const Coin& txout) {
-        ::Serialize(s, VARINT(txout.nHeight * uint32_t{2} + txout.fCoinBase ));
+        ::Serialize(s, VARINT(txout.nHeight * uint32_t{4} + txout.fCoinBase + (txout.fCoinStake ? 2u : 0u)));
+        ::Serialize(s, VARINT(txout.nTime));
         if (txout.nHeight > 0) {
             // Required to maintain compatibility with older undo format.
             ::Serialize(s, (unsigned char)0);
@@ -36,8 +37,10 @@ struct TxInUndoFormatter
     void Unser(Stream &s, Coin& txout) {
         uint32_t nCode = 0;
         ::Unserialize(s, VARINT(nCode));
-        txout.nHeight = nCode >> 1;
+        txout.nHeight = nCode >> 2;
         txout.fCoinBase = nCode & 1;
+        txout.fCoinStake = nCode & 2;
+        ::Unserialize(s, VARINT(txout.nTime));
         if (txout.nHeight > 0) {
             // Old versions stored the version number for the last spend of
             // a transaction's outputs. Non-final spends were indicated with
