diff --git a/src/validationinterface.h b/src/validationinterface.h
index 7c3ce00fbc..e39edbf3e1 100644
--- a/src/validationinterface.h
+++ b/src/validationinterface.h
@@ -109,7 +109,6 @@ protected:
      * - SIZELIMIT (removed in size limiting if the mempool exceeds -maxmempool megabytes)
      * - REORG (removed during a reorg)
      * - CONFLICT (removed because it conflicts with in-block transaction)
-     * - REPLACED (removed due to RBF replacement)
      *
      * This does not fire for transactions that are removed from the mempool
      * because they have been included in a block. Any client that is interested
