diff --git a/src/node/context.h b/src/node/context.h
index 644c997531..e3ad491618 100644
--- a/src/node/context.h
+++ b/src/node/context.h
@@ -10,21 +10,23 @@
 #include <memory>
 #include <vector>
 
+#include <interfaces/init.h>
+#include <interfaces/chain.h>
+#include <interfaces/wallet.h>
+
 class ArgsManager;
 class BanMan;
 class AddrMan;
-class CBlockPolicyEstimator;
 class CConnman;
 class CScheduler;
 class CTxMemPool;
 class ChainstateManager;
 class PeerManager;
-namespace interfaces {
-class Chain;
-class ChainClient;
-class Init;
-class WalletLoader;
-} // namespace interfaces
+
+using interfaces::Chain;
+using interfaces::ChainClient;
+using interfaces::Init;
+using interfaces::WalletLoader;
 
 namespace node {
 //! NodeContext struct containing references to chain state and connection
@@ -43,7 +45,6 @@ struct NodeContext {
     std::unique_ptr<AddrMan> addrman;
     std::unique_ptr<CConnman> connman;
     std::unique_ptr<CTxMemPool> mempool;
-    std::unique_ptr<CBlockPolicyEstimator> fee_estimator;
     std::unique_ptr<PeerManager> peerman;
     std::unique_ptr<ChainstateManager> chainman;
     std::unique_ptr<BanMan> banman;
